### ISSUE TITLE
change dataset, 

### ISSUE THREAD

**Issue by LU4E** [facebookresearch/maskrcnn-benchmark#77]
## ❓ Questions and Help
I change the dataset, which i want to focus on segmentation. I offer the mask and label, but the code :
maskrcnn_benchmark.modeling.matcher.Matcher have the problem. The matcher need the N*M tensor.
but the boxlist_iou(target, proposal) output shape is:
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 126])
boxlist ops.py: iou shape is torch.Size([1, 128])
boxlist ops.py: iou shape is torch.Size([1, 132])
boxlist ops.py: iou shape is torch.Size([1, 127])
boxlist ops.py: iou shape is torch.Size([1, 126])
boxlist ops.py: iou shape is torch.Size([1, 125])
boxlist ops.py: iou shape is torch.Size([1, 130])
boxlist ops.py: iou shape is torch.Size([1, 123])
boxlist ops.py: iou shape is torch.Size([1, 128])
boxlist ops.py: iou shape is torch.Size([1, 133])
boxlist ops.py: iou shape is torch.Size([1, 119])
boxlist ops.py: iou shape is torch.Size([1, 126])
boxlist ops.py: iou shape is torch.Size([1, 121])
boxlist ops.py: iou shape is torch.Size([1, 135])
boxlist ops.py: iou shape is torch.Size([1, 108])
boxlist ops.py: iou shape is torch.Size([1, 129])
boxlist ops.py: iou shape is torch.Size([1, 1])

the result is 

![47758591-9e3a2680-dce6-11e8-88b8-4cb2e03e423e](https://user-images.githubusercontent.com/12681953/47785943-3109ae00-dd45-11e8-842e-bf7a685b2601.png)
![47758594-a2feda80-dce6-11e8-80db-dd848a0e60af](https://user-images.githubusercontent.com/12681953/47785952-36ff8f00-dd45-11e8-9548-7fed6c0ef664.png)
![47758596-a4c89e00-dce6-11e8-9755-e82564d82174](https://user-images.githubusercontent.com/12681953/47785956-3b2bac80-dd45-11e8-8db9-f96a31ac66d6.png)
![47758598-a72af800-dce6-11e8-9be4-36ff33b0d1ca](https://user-images.githubusercontent.com/12681953/47785962-3e269d00-dd45-11e8-8548-29a76b307089.png)

I have some questions:
1. why this boxlist_iou run 33 times, my config size_divisibility is 16.
2. why  [1,1] will let the segmentation_masks become [0]
3. How can i run the code step by step, which could help me unstand all of the code?

**Comment by fmassa** [comment]
Hi,

If you are training on 8 GPUs, it's normal that it gets printed many times, because each GPU will print it once.
In this case, see that the `boxlist ops.py: iou shape is torch.Size([1, 16368])` is printed 16 times, which seems normal for 2 images per GPU, 8 GPUs.

To run the code step-by-step, do not use multiple GPUs, run using a single GPU without `-m torch.distributed.launch`, and debug from there.

About the rest of the question, I unfortunately couldn't understand the issue very well, maybe you could try explaining a bit more?

**Comment by LU4E** [comment]
I am training on single GPU, and i try the gdb. But it dosen't help, i am still studying how to use it.
The problem is: the mask will be [0] after
def match_targets_to_proposals(self, proposal, target):
        match_quality_matrix = boxlist_iou(target, proposal)
        matched_idxs = self.proposal_matcher(match_quality_matrix)
        # Mask RCNN needs "labels" and "masks "fields for creating the targets
        target = target.copy_with_fields(["labels", "masks"])
        # get the targets corresponding GT for each proposal
        # NB: need to clamp the indices because we can have a single
        # GT in the image, and matched_idxs can be -2, which goes
        # out of bounds
        matched_targets = target[matched_idxs.clamp(min=0)]
        matched_targets.add_field("matched_idxs", matched_idxs)
        return matched_targets

this block.
I don't figure out the shape of matched_idxs should be.

**Comment by fmassa** [comment]
If I understand it correctly, the reason might be that you have one (or more) images in your dataset that do not have boxes.

Can you try https://github.com/facebookresearch/maskrcnn-benchmark/pull/37 and follow the discussion in https://github.com/facebookresearch/maskrcnn-benchmark/issues/31 to see if that's indeed the case?

**Comment by LU4E** [comment]
i build the box manually with the bounder of the mask, and every sample should have one box at least. But in some sample ,the box could be every every small.
i will check the box infomation and the code again tomorrow, because the lab is closing.
thank u very very very much for ur reply so many times.

**Comment by LU4E** [comment]
i have check all of the issue, and i also find the issue31. But i don't think my problem is the same one with that. Because i kick out the sample without the mask before i build the dataset, and the box is build according to the mask. I am sure every sample should have one box.

**Comment by fmassa** [comment]
Could you try applying the patch from #37 in your code so that we have an idea if the problem is there or not?

I'm having troubles understanding what else the problem is actually.

**Comment by LU4E** [comment]
i change the code like this:

![2018-11-01 07-56-25](https://user-images.githubusercontent.com/12681953/47825227-a8722880-ddab-11e8-885f-dd12b361af18.png)

unfortunately it doesn't work. there is no error be caught.

and i also print the boxes infomation in the function __getitem__, it give the result like this:

2018-11-01 07:54:12,314 maskrcnn_benchmark.trainer INFO: Start training
dataset.py boxes is [[149 100 180 130]]
dataset.py boxes is [[ 75  95 122 189]]
dataset.py boxes is [[100 167 124 186]]
dataset.py boxes is [[ 77 109  96 145]]
dataset.py boxes is [[ 64  98 116 175]]
dataset.py boxes is [[162 114 190 151]]
dataset.py boxes is [[ 68 120 119 198]]
dataset.py boxes is [[ 69  94 120 167]]
dataset.py boxes is [[104  91 124 135]]
dataset.py boxes is [[ 78 154 111 188]]
dataset.py boxes is [[ 82  84 130 146]]
dataset.py boxes is [[ 78 102 125 176]]
dataset.py boxes is [[153 136 175 199]]
dataset.py boxes is [[ 90  70 129 148]]
dataset.py boxes is [[147 138 191 202]]
dataset.py boxes is [[134  89 189 176]]
dataset.py boxes is [[161 170 168 182]]
dataset.py boxes is [[ 60 113 118 220]]
dataset.py boxes is [[ 81  58 144 126]]
dataset.py boxes is [[ 59  96 127 203]]
dataset.py boxes is [[126  85 179 123]]
dataset.py boxes is [[142  97 171 155]]
dataset.py boxes is [[ 82 101 106 120]]
dataset.py boxes is [[138 131 142 137]]
dataset.py boxes is [[126 157 176 214]]
dataset.py boxes is [[133 139 183 182]]
dataset.py boxes is [[163 111 193 142]]
dataset.py boxes is [[ 66 123 127 192]]
dataset.py boxes is [[125 144 172 200]]
dataset.py boxes is [[118 102 160 149]]
dataset.py boxes is [[113  61 182 135]]
dataset.py boxes is [[130  83 169 153]]
dataset.py boxes is [[ 88 157 108 189]]
dataset.py boxes is [[159 105 189 136]]
dataset.py boxes is [[116  82 172 172]]
dataset.py boxes is [[ 79  96 108 133]]
dataset.py boxes is [[160 141 195 196]]
dataset.py boxes is [[142  99 155 131]]
dataset.py boxes is [[ 88 151 125 209]]
dataset.py boxes is [[132 166 184 200]]
dataset.py boxes is [[ 86 112 118 197]]
dataset.py boxes is [[ 87 153 103 193]]
dataset.py boxes is [[ 84 145 115 177]]
dataset.py boxes is [[131 146 178 214]]
dataset.py boxes is [[ 80  85 125 137]]
dataset.py boxes is [[ 77  75 130 157]]
dataset.py boxes is [[122 143 171 208]]
dataset.py boxes is [[102 132 122 160]]
dataset.py boxes is [[151 134 191 169]]
dataset.py boxes is [[143 124 179 171]]
dataset.py boxes is [[ 64  65 136 178]]
dataset.py boxes is [[125 151 170 210]]
dataset.py boxes is [[122  71 166 134]]
dataset.py boxes is [[132  84 181 145]]
dataset.py boxes is [[ 84 168 148 211]]
dataset.py boxes is [[ 87 106  93 122]]
dataset.py boxes is [[ 68 119 114 217]]
dataset.py boxes is [[103 134 122 159]]
dataset.py boxes is [[147 114 190 190]]
dataset.py boxes is [[130 126 146 152]]
dataset.py boxes is [[ 71  63 136 156]]
dataset.py boxes is [[ 61  98  95 146]]
dataset.py boxes is [[ 67 133 114 205]]
dataset.py boxes is [[143  93 195 149]]
dataset.py boxes is [[ 63  69 138 176]]
dataset.py boxes is [[ 60 125 111 201]]
dataset.py boxes is [[160 103 177 132]]
dataset.py boxes is [[112 101 143 127]]
dataset.py boxes is [[131 121 178 188]]
dataset.py boxes is [[140 105 190 173]]
dataset.py boxes is [[ 81 107 119 207]]
dataset.py boxes is [[ 89 129 112 209]]
dataset.py boxes is [[ 60 122 113 198]]
dataset.py boxes is [[ 76 132 133 194]]
dataset.py boxes is [[117  85 192 186]]
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
dataset.py boxes is [[152 102 173 127]]
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
dataset.py boxes is [[145  98 186 140]]
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
boxlist ops.py: iou shape is torch.Size([1, 16368])
dataset.py boxes is [[148 154 175 178]]
boxlist ops.py: iou shape is torch.Size([1, 139])
boxlist ops.py: iou shape is torch.Size([1, 137])
boxlist ops.py: iou shape is torch.Size([1, 147])
boxlist ops.py: iou shape is torch.Size([1, 96])
boxlist ops.py: iou shape is torch.Size([1, 130])
boxlist ops.py: iou shape is torch.Size([1, 130])
boxlist ops.py: iou shape is torch.Size([1, 123])
boxlist ops.py: iou shape is torch.Size([1, 115])
boxlist ops.py: iou shape is torch.Size([1, 124])
boxlist ops.py: iou shape is torch.Size([1, 151])
boxlist ops.py: iou shape is torch.Size([1, 115])
boxlist ops.py: iou shape is torch.Size([1, 126])
boxlist ops.py: iou shape is torch.Size([1, 114])
boxlist ops.py: iou shape is torch.Size([1, 109])
boxlist ops.py: iou shape is torch.Size([1, 133])
boxlist ops.py: iou shape is torch.Size([1, 127])
dataset.py boxes is [[141 101 191 171]]
dataset.py boxes is [[ 66 111 119 207]]
boxlist ops.py: iou shape is torch.Size([1, 1])
dataset.py boxes is [[127 162 148 185]]
dataset.py boxes is [[130 122 168 187]]
dataset.py boxes is [[140  87 165 112]]
dataset.py boxes is [[127  78 179 173]]
dataset.py boxes is [[ 75 102 139 174]]
dataset.py boxes is [[140  91 178 138]]
dataset.py boxes is [[155 139 191 193]]
dataset.py boxes is [[ 86 135 115 165]]
dataset.py boxes is [[134  97 190 197]]
dataset.py boxes is [[ 67  94 116 169]]
dataset.py boxes is [[142 204 145 208]]
dataset.py boxes is [[112  61 173 134]]
dataset.py boxes is [[ 87 149 125 209]]
dataset.py boxes is [[136  96 155 157]]
dataset.py boxes is [[127 164 154 209]]
dataset.py boxes is [[123 111 134 125]]
dataset.py boxes is [[ 63 115 118 213]]
dataset.py boxes is [[128 116 158 159]]
dataset.py boxes is [[139 197 145 202]]
dataset.py boxes is [[ 85  78 131 152]]
dataset.py boxes is [[114  71 173 130]]
dataset.py boxes is [[138 156 170 211]]
dataset.py boxes is [[135 161 165 220]]
dataset.py boxes is [[ 60  95 137 204]]
dataset.py boxes is [[ 80 141 128 199]]
dataset.py boxes is [[ 78 106  92 123]]
dataset.py boxes is [[ 84 150 126 199]]
dataset.py boxes is [[150 100 178 130]]
dataset.py boxes is [[117  78 163 128]]
dataset.py boxes is [[146 151 148 154]]
dataset.py boxes is [[142  90 190 170]]
dataset.py boxes is [[141 148 175 188]]

the bug is still this one:

Traceback (most recent call last):
AssertionError: [0], BoxList(num_boxes=1, image_width=256, image_height=256, mode=xyxy)

**Comment by LU4E** [comment]
I find the source of problem: the shape of mask. And this issue is similar to the issue #16  .
i don't know what shape the mask of coco offer, my mask shape is [256, 256, 1]

![2018-11-01 10-01-04](https://user-images.githubusercontent.com/12681953/47829031-6d2e2480-ddc0-11e8-9eae-6251cddadb5f.png)

that leads to the matched_target get a wrong mask:

![2018-11-01 10-01-15](https://user-images.githubusercontent.com/12681953/47829034-715a4200-ddc0-11e8-812d-531f738183fb.png)

I change the format of mask to [1, 256, 256], the target recive the mask as [1, 256, 256], the matched_targets get the mask [256, 256],

![2018-11-01 10-27-52](https://user-images.githubusercontent.com/12681953/47829134-f34a6b00-ddc0-11e8-86e8-321c4914a7fb.png)


 that will have another problem:
after this code positive_proposals = proposals_per_image[positive_inds] :

![2018-11-01 10-29-46](https://user-images.githubusercontent.com/12681953/47829308-b7fc6c00-ddc1-11e8-9bf8-0c9837c11e50.png)

so i should change the code or i change the mask format.
could you explain the mask shape should be in this model?

**Comment by fmassa** [comment]
Oh, now I see the issue I think

`SegmentationMask` currently assumes the masks are represented as a set of polygons, and not a binary segmentation mask.

This can be modified by adding support for binary segmentation masks as well.
This is almost fully implemented, but as I didn't have any dataset to test it, I didn't finish it up.

Have a look at `segmentation_mask.py` and the `Mask` class there to see what I mean.
I suppose this is the problem maybe?

**Comment by LU4E** [comment]
what should i do now?
the mask must turn to be represented as a set of polygons for now?

**Comment by fmassa** [comment]
For bow, you can also try modifying `SegmentationMask` to support `Mask` object instead of `Polygon`. It should not be hard, but requires some coding. Let me know if you require more guidance and I can help you.

**Comment by LU4E** [comment]
Yes, I need the help. Now the biggest problem is that i don't know the mask format needed by model.
So could you tell me  what function the class Mask should offer? Dose it need to adjust  the class SegmentationMask ?

**Comment by fmassa** [comment]
@LU4E have a look at [Mask](https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/maskrcnn_benchmark/structures/segmentation_mask.py#L11-L48). This starts the implementation of what you need, which is a class that holds segmentation masks as 2d tensors. You'll need to implement the `resize` method there (which can be done with `torch.nn.functional.interpolate`), and then modify the `SegmentationMask` class to use a [list of `Mask` instead of a list of `Polygons`](https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/maskrcnn_benchmark/structures/segmentation_mask.py#L161).

One last thing, you might also need to implement a `convert` method in `Mask`, which simply returns the underlying `mask` data as a `Tensor`.

Basically, if you make the API of `Mask` be the same as the API in `Polygons`, and replace
```python
self.polygons = [Polygons(p, size, mode) for p in polygons]
```
by
```python
self.polygons = [Mask(p, size, mode) for p in polygons]
```
in [here](https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/maskrcnn_benchmark/structures/segmentation_mask.py#L161), it should be enough for the code to run.

**Comment by LU4E** [comment]
OK，let me have a try.

**Comment by LU4E** [comment]
![2018-11-02 18-51-28](https://user-images.githubusercontent.com/12681953/47911552-b53c6c80-ded0-11e8-9d47-d69f8a80b715.png)
![2018-11-02 18-52-03](https://user-images.githubusercontent.com/12681953/47911565-bb324d80-ded0-11e8-8ed8-74a5749627da.png)
![2018-11-02 18-50-21](https://user-images.githubusercontent.com/12681953/47911568-bec5d480-ded0-11e8-8690-3f2ab9f28182.png)
![2018-11-02 18-54-00](https://user-images.githubusercontent.com/12681953/47911570-c2595b80-ded0-11e8-9f94-fe01bb7064e2.png)
there is a bug in crop:
![2018-11-02 19-00-39](https://user-images.githubusercontent.com/12681953/47911865-9db1b380-ded1-11e8-806b-39fa9c724f42.png)

![2018-11-02 19-00-12](https://user-images.githubusercontent.com/12681953/47911872-a30efe00-ded1-11e8-9489-4153878665f1.png)

**Comment by fmassa** [comment]
There is a problem with your implementation I think: the `Mask` class is supposed to hold a single 2d segmentation mask (with possibly many channels), so `self.masks` should be a 3d Tensor, and the `resize` function shouldn't be passing a list to `Mask`.

**Comment by LU4E** [comment]
The class SegmentationMask have assert the input as a list. So I turn mask to be a list to keep the code working for both of mask and polygon. I am not familiar with the polygon. If it's not necessary to ensure the masks as a list, can I just delete the code of assert  isinstance(ploy on,list)?

**Comment by fmassa** [comment]
So, you should think of `SegmentationMask` as a list of `Mask`, and a `Mask` as a single Tensor,
But I believe that in your case you made `scaled_masks` be a list.

**Comment by LU4E** [comment]
I understand your point. I thought the segmentationmask receive a list of polygon,and dived it to each instance automatically, So as to mask, that is why I didn't change the segmentationmask. I will adjust the code of Mask. And Google the polygon to understand all of this class. Because I am working on another job, which costs too much time, resulting in having no time to check the flow of the data. I am appreciate your help very very much. And thank you again.

**Comment by fmassa** [comment]
No worries!

That being said, it would be great to have full support for `Mask` in the codebase, so PRs adding support for that would be more than welcome!

**Comment by LU4E** [comment]
Yes. At least for me, I'm desperate for this feature.
Because It's  clear to me that what I have done to the array, but I have no idea how to processing the polygon.

**Comment by wangg12** [comment]
@fmassa +1 for the support of this feature and the need for the feature of converting RLE/polygons format to binary mask.

**Comment by fmassa** [comment]
@wangg12 I agree. The support for `Mask` is almost all already done by @LU4E , support for converting to binary mask can be added at a later step.

**Comment by LU4E** [comment]
@Spandan-Madan 
I don't know what the data looks like, could you offer the dataloader which you build?
there are two suggestion:
first: the image shape should be NxCxHxW, pytorch is channel first.
second: have you add the field to the target in the dataloader, which can be done like this: targets.add_field('mask',mask)

**Comment by LU4E** [comment]
the targets have fields like:mask, label. Build a boxlist, and add the field to it.
If you don't want to  change the code of the mask-rcnn, make sure your dataset have the same format as the sample it offers.


# create a BoxList from the boxes
boxlist = BoxList(boxes, image.size, mode="xyxy")
# add the labels to the boxlist
boxlist.add_field("labels", labels)

you can write your code like this.

**Comment by fmassa** [comment]
@Spandan-Madan sorry if parts of the code are confusing.
Indeed, detection models are very complicated, and its behavior during training depends on the ground-truth, which makes things even more complicated. The reason is that the `RPN` uses the ground-truths during training to select which boxes to feed to the detection heads, so only following a `output = model(inputs); loss = criterion(output, target)` is not going to be as efficient / easy.
But during inference you should be able to simply do `model(inputs)`.

About segmentation masks, do you want to do instance segmentation or semantic segmentation?
If you want to do instance segmentation, for each mask you can compute its bounding box during `__getitem__` of your model, and attach it to the `BoxList`.

If you want to do semantic segmentation (where instances are not necessary), then this codebase is not yet adapted to your needs.

Let us know if you have further questions.

**Comment by LU4E** [comment]
no, i think you make a mistake.
this box is just for the detection, the label for your segmentation is the mask.

**Comment by Spandan-Madan** [comment]
@fmassa That makes sense. We don't have instance level labels yet, so we can't get a bounding box around each object.
